tests/legacy/test_resource_warnings.py

```diff
@@ -17,10 +19,21 @@ def test_task_double_close(self, init_kwargs):
             with nidaqmx.Task(**init_kwargs) as task:
                 task.close()
 
+    @pytest.mark.library_only
     def test_unclosed_task(self, init_kwargs):
         """Test to validate unclosed tasks."""
         task = nidaqmx.Task(**init_kwargs)
         # Since __del__ is not guaranteed to be called, for the purposes of
         # consistent test results call __del__ manually.
         with pytest.warns(DaqResourceWarning):
             task.__del__()
```

Comment:
BTW, pytest reports a warning from this test:

```
================================================== warnings summary ===================================================
tests/legacy/test_resource_warnings.py::TestResourceWarnings::test_unclosed_task[grpc_init_kwargs]
  D:\dev\nidaqmx-python\generated\nidaqmx\task.py:93: ResourceWarning: Task of name "_unnamedTask<0>" was not explicitly closed before it was destructed. Resources on the task device may still be reserved.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================== 2 passed, 745 deselected, 1 warning in 2.46s =====================================
```

I don't think `task.__del__()` actually destroys the task object. It's just a callback that the GC invokes before collecting the object.

Reply:
I updated the tests but there is one discrepancy found, I performed task leaking on both library task and grpc task, expected library interpreter task warns but grpc task doesn't warn. Tests passed as expected. There is one warning by GC from library interpreter task, but in test summary, the warning is categorized in grpc task test method
```
import warnings
import pytest

import nidaqmx
from nidaqmx import DaqResourceWarning

@pytest.mark.library_only
def test__library_task__leak_task__raised_resource_warning(init_kwargs):
    task = nidaqmx.Task("LibraryTask", **init_kwargs)

    with pytest.warns(DaqResourceWarning):
        task.__del__()
    
    # Closing the task resolves the warning.
    # task.close()

@pytest.mark.grpc_only
def test__grpc_task__leak_task__not_raised_resource_warning(init_kwargs):
    task = nidaqmx.Task("GrpcTask", **init_kwargs)

    with warnings.catch_warnings(record=True) as warnings_raised:
        task.__del__()

    assert len(warnings_raised) == 0
```

```
====================================================== warnings summary ====================================================== 
tests/component/test_sample_task_res_warns.py::test__grpc_task__leak_task__not_raised_resource_warning[grpc_init_kwargs]       
  C:\Users\battu.mounika\source\repos\TAF\NI-DAQmxPython New\nidaqmx-python\generated\nidaqmx\task.py:94: ResourceWarning: Task of name "LibraryTask" was not explicitly closed before it was destructed. Resources on the task device may still be reserved. 
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================ 2 passed, 1 warning in 3.28s ================================================
```

Closing the Library interpreter task resolves the warning. But not sure why the library interpreter warning is caught in grpc task test method. Please share your thoughts?

Reply:
I added a Test case `test__closed_task__del_task__resource_warning_not_raised` in `test_task_resource_warnings.py` to validate `__del__` of a closed task.

Reply:
If you call `task.__del__()` and then leak the task, the GC calls `task.__del__()` again outside the `with pytest.warns()` block.

`test__closed_task__del_task__resource_warning_not_raised` looks fine.